File: 25% Programming Assignment 2-20181017/25% Programming Assignment 2-20181022/JavaScript_Mitre10Cup/JavaScript_Mitre10Cup/src/Competition.js
/* globals Team, Game, View */
class Competition { // eslint-disable-line no-unused-lets
  constructor (newSponsor, newYear) {
    this.sponsor = newSponsor
    this.year = newYear
    this.allMyChampionshipTeams = []
    this.allMyPremiershipTeams = []
    this.allMyTeams = []
    this.allMyGames = []
  }
  
  addTeam (newRank, newName, newHome, newCity) {
    let aNewTeam = new Team(newRank, newName, newHome, newCity)
    this.allMyTeams.push(aNewTeam)
    if (newRank <= 7) {
      this.allMyPremiershipTeams.push(aNewTeam)
    } else {
      this.allMyChampionshipTeams.push(aNewTeam)
    }
  }
  
  findTeamByRank (targetRank) {
    return this.allMyTeams.find((team) => { return team.rank === targetRank })
  }
  
  addGame (newWeek, newHomeTeamRank, newAwayTeamRank, newWhenString) { // newYear, newMonth, newDay, newMinute) {
    // let when = new Date(newYear, newMonth, newDay, newMinute)
    let when = new Date(newWhenString)
    let theHomeTeam = this.findTeamByRank(newHomeTeamRank)
    let theAwayTeam = this.findTeamByRank(newAwayTeamRank)
    let aNewGame = new Game(newWeek, theHomeTeam, theAwayTeam, when)
    this.allMyGames.push(aNewGame)
  }
  
  findGameByRank (homeRank, awayRank){
    //return this.allMyGames.find((game) => {return (game.homeTeamRank == homeRank)&&(game.awayTeamRank == awayRank)})
	let result = null
	for (let aGame of this.allMyGames){
		let atr = aGame.myAwayTeam.rank
		let htr = aGame.myHomeTeam.rank
		if ((atr == awayRank) && (htr == homeRank)){
			result = aGame
			return result
		}
	}	
  }
  setResult (homeTeamRank, homeTeamScore, homeTeamTries, awayTeamRank, awayTeamScore, awayTeamTries) { // newYear, newMonth, newDay, newMinute) {
    // let when = new Date(newYear, newMonth, newDay, newMinute)
    let aGame = this.findGameByRank(homeTeamRank, awayTeamRank)
    aGame.setResult(homeTeamScore,awayTeamScore)
    aGame.homeTeamTries = homeTeamTries
    aGame.awayTeamTries = awayTeamTries
    
    
    
    /*let when = new Date(newWhenString)
    let theHomeTeam = this.findTeamByRank(newHomeTeamRank)
    let theAwayTeam = this.findTeamByRank(newAwayTeamRank)
    let aNewGame = new Game(newWeek, theHomeTeam, theAwayTeam, when)
    this.allMyGames.push(aNewGame)
    */
  }
  
  getDivisions () {
    let result = `${View.NEWLINE()}TEAMS${View.NEWLINE()}Premiership Division${View.NEWLINE()}`
    for (let aTeam of this.allMyPremiershipTeams) {
      result += `${aTeam.rank} ${aTeam.name} ${View.NEWLINE()}`
    }
    result += 'Championship Division' + View.NEWLINE()
    for (let aTeam of this.allMyChampionshipTeams) {
      result += `${aTeam.rank} ${aTeam.name} ${View.NEWLINE()}`
    }
    return result
  }
  
  getGames () {
    let result = `${View.NEWLINE()}GAMES${View.NEWLINE()}`
    let week = 0
    for (let aGame of this.allMyGames) {
      if (aGame.week !== week) {
        week += 1
        result += `Week ${week}. ${aGame.getWhen()} - ${aGame.getSunday()}${View.NEWLINE()}`
      }
      result += aGame + View.NEWLINE()
    }
    return result
  }
  
  getCanterburyGames () {
    let result = `${View.NEWLINE()}CANTERBURY GAMES${View.NEWLINE()}`
    for (let aGame of this.allMyGames) {
      if (aGame.hasTeam('Canterbury')) {
        result += aGame.get() + View.NEWLINE()
      }
    }
    return result
  }

  getCrossOverGames () {
    let result = `${View.NEWLINE()}CROSS OVER GAMES${View.NEWLINE()}`
    for (let aGame of this.allMyGames) {
      if (aGame.isCrossOver()) {
        result += aGame.get() + View.NEWLINE()
      }
    }
    return result
  }
  
  ///NEW FUNCTIONS FROM ASSIGNMENT 2 FOR DISPLAYING THE STANDINGS
  
  getStandings(){
    let result = ""
    for(let aTeam of this.allMyTeams){
      this.setStandings(aTeam)
    }
    this.sortStandings(this.allMyPremiershipTeams)
	  result += `${View.NEWLINE()}STANDINGS ${View.NEWLINE()}${this.sponsor} PREMIERSHIP ${View.NEWLINE()}`
	  for(let aTeam of this.allMyPremiershipTeams){
      
      result += `${View.NEWLINE()}${aTeam.name}${aTeam.get()}${View.NEWLINE()}`
    }
    this.sortStandings(this.allMyChampionshipTeams)
    result +=`${View.NEWLINE()}${View.NEWLINE()}${this.sponsor} CHAMPIONSHIP ${View.NEWLINE()}`
    for(let aTeam of this.allMyChampionshipTeams){
      result += `${View.NEWLINE()}${aTeam.name}${View.NEWLINE()}${aTeam.get()}${View.NEWLINE()}`
    }
    return result
  }
  
  //let result = `${View.NEWLINE()}<b>${aTeam.name}:</b>`
  setStandings(aTeam){ 
    for(let aGame of this.allMyGames){
      if(aGame.hasTeam(aTeam.name)){
        aTeam.standings(aGame)
      }
    }
  }
  

  sortStandings(teamList){
    teamList.sort(function(a, b){return a.gamesWon - b.gamesWon})
    teamList.reverse()
  }
  
  
  getAll () {
    let result = this.getDivisions()
    result += this.getGames()
    result += this.getCanterburyGames()
    result += this.getCrossOverGames()
    result += this.getStandings()
    return result
  }
}

File: 25% Programming Assignment 2-20181017/25% Programming Assignment 2-20181022/python_mitre10_cup/python_mitre10_cup/competition.py
from team import Team
from game import Game
import datetime

class Competition(object):
    def __init__(self, new_sponsor, new_year):
        self.sponsor = new_sponsor
        self.year = new_year
        self.all_premiership_teams = []
        self.all_championship_teams = []
        self.all_games = []
        self.all_teams = []
        self.teamCount = 0  # used to separate Premiership and Championship teams by rank

    def add_team(self, new_rank, new_name, new_venue, new_city):
        self.teamCount += 1
        a_new_team = Team(new_rank, new_name, new_venue, new_city)
        self.all_teams.append(a_new_team)
        if self.teamCount <= 7:
            self.all_premiership_teams.append(a_new_team)
        else:
            self.all_championship_teams.append(a_new_team)

    def get_team_by_rank(self, target_rank):
        return self.all_teams[target_rank]

    def add_game(self, new_round, new_home_team_rank, new_away_team_rank, new_when_is8601string):
        import dateutil.parser
        when = dateutil.parser.parse(new_when_is8601string)
        home_team = self.get_team_by_rank(new_home_team_rank - 1)
        away_team = self.get_team_by_rank(new_away_team_rank - 1)
        a_new_game = Game(new_round, home_team, away_team, when)
        self.all_games.append(a_new_game)

    def get_divisions(self):
        result = '\nTEAMS\nPremiership Division\n'
        for team in self.all_premiership_teams:
            result += f'\t{team.rank} {team.name}\n'
        result += 'Championship Division\n'
        for team in self.all_championship_teams:
            result += f'\t{team.rank} {team.name}\n'
        return result

    def get_games(self):
        result = '\nGAMES\n'
        week = 0
        for a_game in self.all_games:
            if a_game.week != week:
                week += 1
                pretty_first_game_date = a_game.when.strftime('%A %d %B')
                sunday_date = a_game.when + datetime.timedelta(6-a_game.when.weekday())
                pretty_sunday_date = sunday_date.strftime('%A %d %B')
                result += f'Week {week}. {pretty_first_game_date} - {pretty_sunday_date}\n'
            result += str(a_game) + '\n'
        return result

    def get_canterbury_games(self):
        result = 'CANTERBURY GAMES\n'
        for a_game in self.all_games:
            if a_game.has_team('Canterbury'):
                result += a_game.get() + '\n'
        return result

    def get_cross_over_games(self):
        result = '\nCROSS OVER GAMES\n'
        for a_game in self.all_games:
            if a_game.is_crossover():
                result += a_game.get() + '\n'
        return result

    def __str__(self):
        result = self.get_divisions()
        result += self.get_games()
        result += self.get_canterbury_games()
        result += self.get_cross_over_games()
        return result

    def get(self):
        result = []
        for g in self.all_games:
            result.append( g.when.isformat(), g.home_team.rank,  g.away_team.rank)

    def getResults(self):
        result =